---
title: "Comparison of Viridian and Genbank assemblies"
subtitle: Bethany Dearlove, Lukas Endler and Andreas Bergthaler
author: | 
  | Institute for Hygiene and Applied Immunology, Medical University of Vienna</br>
  | Contact: bethany.dearlove@meduniwien.ac.at, andreas.bergthaler@meduniwien.ac.at
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE,message=FALSE, warning=FALSE, cache=TRUE,cache.lazy = FALSE, dpi=100,fig.align = "center")

library("tidyverse")
library("scico")
library("scales")
library("lubridate")
library("patchwork")
library("jsonlite")
library("DT")
library("GenomicFeatures")
library("VariantAnnotation")
library("genbankr")
library("htmltools")
library("conflicted")
library("ggrepel")
library("ggtext")

conflicts_prefer(
  dplyr::select,
  dplyr::rename,
  dplyr::count,
  dplyr::filter,
  stringr::fixed,
  base::setdiff,
  .quiet=TRUE
)

#Inline code output formatting of numbers
#Round to 2 decimal places and add commas for large numbers
knitr::knit_hooks$set(inline = function(x) {
  if(is.numeric(x)&str_detect(x,"\\.")){
      x <- format(round(x,2),big.mark=",",nsmall = 2)
  }else{
    x <- format(x,big.mark=",")
  }
})

#Code to be able to specify figure sizes in asis loop
#Adapted from https://stackoverflow.com/questions/15365829/dynamic-height-and-width-for-knitr-plots
subchunkify <- function(g, chunk_name=NULL, fig_height=7, fig_width=5) {
  g_deparsed <- paste0(deparse(
    function() {g}
  ), collapse = '')

  if(is.null(chunk_name)){
      sub_chunk <- paste0("
  `","``{r subchunk_", floor(runif(1) * 10000), ", fig.height=",
   fig_height, ", fig.width=", fig_width, ", echo=FALSE}",
  "\n(", 
    g_deparsed
    , ")()",
  "\n`","``
  ")
  }else{
      sub_chunk <- paste0("
  `","``{r subchunk_", chunk_name, ", fig.height=",
   fig_height, ", fig.width=", fig_width, ", echo=FALSE}",
  "\n(", 
    g_deparsed
    , ")()",
  "\n`","``
  ")
  }

  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}

#Initialise a data table
datatable(cars,extensions="FixedColumns")

```

```{css, eval = TRUE, echo = FALSE}
/*Makes the table of contents indentations a little nicer*/

.tocify-header {
  text-indent: initial;
}

.tocify-subheader > .tocify-item {
  text-indent: initial;
  padding-left: 20px;
}

.tocify-subheader .tocify-subheader > .tocify-item {
  text-indent: initial;
  padding-left: 30px;
}

.tocify-subheader .tocify-subheader .tocify-subheader > .tocify-item {
  text-indent: initial;
  padding-left: 40px;
}

.tocify-subheader .tocify-subheader .tocify-subheader > .tocify-item {
  text-indent: initial;
  padding-left: 50px;
}

.tab > div {
  display:none;  
}

.tab > div:target {
  display:block;  
}
```

```{r setColours}
#colours for mismatch plots
cols_mismatch<-setNames(c("steelblue","firebrick"),c("TRUE","FALSE"))
cols_mismatch_rel<-setNames(c("steelblue","#FEC44F","#FB9A29","#EC7014","#CC4C02","darkgrey"),c("Same","Parent-child","Sibling","Same variant","More distant","One unassigned"))

#colours for Viridian and Genbank comparisons
cols_assembly<-c("Viridian"="#40826D","Genbank"="#fc9c5b")
cols_assembly_dark<-c("Viridian"="#295446","Genbank"="#D95F02")
```

```{r getMetadata}
# Set data directory
data_dir<-"./"

# Read in the metadata
metadata_update_number<-"v04"
metadata<-read_tsv(paste0(data_dir,"run_metadata.v04.tsv.gz"))
full_metadata_length<-dim(metadata)[1]


# Keep only intersection between Genbank and Viridian
metadata<-metadata%>%
  filter(In_intersection==TRUE)%>%
  mutate(Viridian_N=as.numeric(Viridian_N),
         Genbank_N=as.numeric(Genbank_N))
```

```{r unalias}
# Get the alias key from the pango github
aliases<-rjson::fromJSON(file="https://raw.githubusercontent.com/cov-lineages/pango-designation/master/pango_designation/alias_key.json")

# Function to dealias variants
dealias <- function(x){
  base <- strsplit(x, split="\\.")[[1]][1]

  if(!any(names(aliases) == base)){
    y <- base
  } else if(nchar(aliases[names(aliases) == base]) == 0){
    y <- base
  } else if(grepl("^X", base)){
    y <- base
  } else {
    y <- aliases[names(aliases) == base]
  }
  dealiased <- gsub(base, y, x)
  return(dealiased)
}

# Function to get WHO name from unaliased lineage
get_who_name<-function(unaliased){case_when(
    unaliased=="B.1.1.7"~"Alpha",
    str_detect(unaliased,"B\\.1\\.1\\.7\\.")~"Alpha",
    unaliased=="B.1.351"~"Beta",
    str_detect(unaliased,"B\\.1\\.351\\.")~"Beta",
    unaliased=="B.1.1.28.1"~"Gamma",
    str_detect(unaliased,"B\\.1\\.1\\.28\\.1\\.")~"Gamma",
    unaliased=="B.1.617.2"~"Delta",
    str_detect(unaliased,"B\\.1\\.617\\.2\\.")~"Delta",
    unaliased=="B.1.1.529"~"Omicron",
    str_detect(unaliased,"B\\.1\\.1\\.529\\.")~"Omicron",
    unaliased=="B.1.617.1"~"Kappa",
    str_detect(unaliased,"B\\.1\\.617\\.1\\.")~"Kappa",
    unaliased%in%c("B.1.427","B.1.429")~"Epsilon",
    str_detect(unaliased,"B\\.1\\.427\\.")~"Epsilon",
    str_detect(unaliased,"B\\.1\\.429\\.")~"Epsilon",
    unaliased=="B.1.525"~"Eta",
    str_detect(unaliased,"B\\.1\\.525\\.")~"Eta",
    unaliased=="B.1.526"~"Iota",
    str_detect(unaliased,"B\\.1\\.526\\.")~"Iota",
    unaliased=="B.1.621"~"Mu",
    str_detect(unaliased,"B\\.1\\.621\\.")~"Mu",
    unaliased=="B.1.1.1.37"~"Lambda",
    str_detect(unaliased,"B\\.1\\.1\\.1\\.37\\.")~"Lambda",
    unaliased=="B.1.1.28.2"~"Zeta",
    str_detect(unaliased,"B\\.1\\.1\\.28\\.2\\.")~"Zeta",
    unaliased=="B.1.1.28.3"~"Theta",
    str_detect(unaliased,"B\\.1\\.1\\.28\\.3\\.")~"Theta",
    str_detect(unaliased,"^X")~"Recombinant",
    unaliased=="Unassigned"~"Unassigned",
    T~"Other")}

# Get dealiased pango lineages
metadata<-metadata%>%
  rowwise()%>%
  mutate(Viridian_pangolin_unaliased=dealias(Viridian_pangolin),
         Genbank_pangolin_unaliased=dealias(Genbank_pangolin))%>%
  ungroup()

# Get a data frame of all lineages and their aliases in the data
lineages_all<-data.frame(lineage=unique(c(metadata$Viridian_pangolin,metadata$Genbank_pangolin)))%>%
  rowwise()%>%
  mutate(lineage_unaliased=dealias(lineage))%>%
  arrange(str_rank(lineage_unaliased,numeric=T))

# Extract a vector of unique variants in the data
variants_unique<-c(metadata$Viridian_pangolin_unaliased,metadata$Genbank_pangolin_unaliased)%>%
  unique()%>%
  get_who_name()%>%
  unique()%>%
  str_sort()%>%
  fct_relevel("Recombinant","Other","Unassigned",after=Inf)%>%
  levels()

# Add new columns for the variant, ordered lineage designations, whether lineages are mismatched between assemblies, whether variants are mismatched between assemblies, and the relationship between lineages when they're mismatched
metadata<-metadata%>%
  mutate(Viridian_pangolin_variant=get_who_name(Viridian_pangolin_unaliased),
         Genbank_pangolin_variant=get_who_name(Genbank_pangolin_unaliased),
         Viridian_pangolin_variant=factor(Viridian_pangolin_variant,levels=variants_unique),
         Genbank_pangolin_variant=factor(Genbank_pangolin_variant,levels=variants_unique),
         Viridian_pangolin_unaliased=factor(Viridian_pangolin_unaliased,levels=lineages_all$lineage_unaliased),
         Viridian_pangolin_unaliased=fct_relevel(Viridian_pangolin_unaliased,"Unassigned",after=Inf),
         Genbank_pangolin_unaliased=factor(Genbank_pangolin_unaliased,levels=lineages_all$lineage_unaliased),
         Genbank_pangolin_unaliased=fct_relevel(Genbank_pangolin_unaliased,"Unassigned",after=Inf),
         Viridian_pangolin_ordered=factor(Viridian_pangolin,levels=lineages_all$lineage),
         Viridian_pangolin_ordered=fct_relevel(Viridian_pangolin_ordered,"Unassigned",after=Inf),
         Genbank_pangolin_ordered=factor(Genbank_pangolin,levels=lineages_all$lineage),
         Genbank_pangolin_ordered=fct_relevel(Genbank_pangolin_ordered,"Unassigned",after=Inf),
         year_week=floor_date(ymd(Date_tree),"week"),
         variant_same=Viridian_pangolin_variant==Genbank_pangolin_variant,
         variant_same=fct_rev(as.factor(variant_same)),
         pangolin_same=Viridian_pangolin==Genbank_pangolin,
         pangolin_same=fct_rev(as.factor(pangolin_same)),
         Genbank_pangolin_unaliased_parent=word(Genbank_pangolin_unaliased,1,-2,sep=stringr::fixed(".")),
         Viridian_pangolin_unaliased_parent=word(Viridian_pangolin_unaliased,1,-2,sep=stringr::fixed(".")),
         pangolin_mismatch_relationship_full=case_when(
           Genbank_pangolin==Viridian_pangolin~"Same",
           Genbank_pangolin_unaliased==Viridian_pangolin_unaliased_parent~"Viridian Parent",
           Viridian_pangolin_unaliased==Genbank_pangolin_unaliased_parent~"Genbank Parent",
           Genbank_pangolin_unaliased_parent==Viridian_pangolin_unaliased_parent~"Sibling",
           Genbank_pangolin_variant==Viridian_pangolin_variant~"Same variant",
           Genbank_pangolin=="Unassigned"~"Genbank unassigned",
           Viridian_pangolin=="Unassigned"~"Viridian unassigned",
           T~"More distant"
           ),
         pangolin_mismatch_relationship=case_when(
           pangolin_mismatch_relationship_full%in%c("Genbank Parent","Viridian Parent")~"Parent-child",
           pangolin_mismatch_relationship_full%in%c("Genbank unassigned","Viridian unassigned")~"One unassigned",
           T~pangolin_mismatch_relationship_full),
         pangolin_mismatch_relationship=factor(pangolin_mismatch_relationship,levels=c("Same","Parent-child","Sibling","Same variant","More distant","One unassigned"))
         )
```

```{r getIndels}
# Read in the indels file by run (so can cross-match on variant)
indels <- read_json(paste0(data_dir,"all_indels.by_run.json.gz"))

# Read in Cornelius Roemer's pango consensus summary json
consensus<- read_json("https://raw.githubusercontent.com/corneliusroemer/pango-sequences/main/data/pango-consensus-sequences_summary.json")

# Mapping between who name and lineage
voc_parent<-c("Alpha"="B.1.1.7", "Beta"="B.1.351", "Gamma"="P.1", "Delta"="B.1.617.2", "Omicron"="B.1.1.529",
"Kappa"="B.1.617.1","Epsilon"="B.1.427","Epsilon"="B.1.429", "Eta"="B.1.525", "Iota"="B.1.526", "Mu"="B.1.621", "Lambda"="C.37", "Zeta"=
"P.2", "Theta"="P.3")
```

```{r getIndelAnnotations, cache=F}
# Get the gbk file for Wu-Hu-1
gbk <- readGenBank(GBAccession("NC_045512"))

# Get the fasta file for Wu-Hu-1
if(!file.exists(paste0(data_dir,"NC_045512.2.ref.fasta"))){
  names(gbk@sequence)<-"NC_045512.2"
  writeXStringSet(gbk@sequence,paste0(data_dir,"NC_045512.2.ref.fasta"))
}

fafi <- FaFile(paste0(data_dir,"NC_045512.2.ref.fasta"))

if(!file.exists(paste0(data_dir,"NC_045512.2.ref.fasta.fai"))){
  open(fafi)  
}


# Rename ids to get nicer gene names
gbk@transcripts$gene_id <- gbk@transcripts$gene
gbk@cds$gene_id <- gbk@cds$gene

# make the taxa database
txdb <- makeTxDbFromGenBank(gbk)

# set the right sequence name 
seqlevels(txdb) <- "NC_045512.2"
genome(txdb) <- "Wuhan-Hu-1"

# function that takes the an AA change and tries to make it more readable (used in annotate_indel below)
clean_AA_change<-function(REFAA,VARAA,AA_START){
  
  if(is.na(REFAA)|(!str_detect(REFAA,"[[:alpha:]]+"))|
     is.na(VARAA)|(!str_detect(VARAA,"[[:alpha:]]+"))|
     is.na(AA_START)|(!str_detect(AA_START,"[[:digit:]]+"))){
    return(NA)
  }
  
  ref_chars<-str_split_1(REFAA,"")
  var_chars<-str_split_1(VARAA,"")
  
  min_length<-min(length(ref_chars),length(var_chars))
  matching_chars<-sum(ref_chars[1:min_length]==var_chars[1:min_length])
  
  if(matching_chars>0){
    new_REFAA<-str_sub(REFAA,matching_chars+1,-1)
    
    new_VARAA<-str_sub(VARAA,matching_chars+1,-1)
    
    new_AAstart<-case_when(
      new_VARAA==""&nchar(new_REFAA)==1~as.character(AA_START+matching_chars),
      new_VARAA==""~paste(AA_START+matching_chars,AA_START+length(ref_chars)-matching_chars,sep="-"),
      new_REFAA==""&matching_chars==1~as.character(AA_START+matching_chars),
      new_REFAA==""~as.character(AA_START+matching_chars-1),
      T~as.character(AA_START+matching_chars))
      
    out<-case_when(new_REFAA==""~paste0("ins",new_AAstart,new_VARAA),
                   new_VARAA==""~paste0(new_REFAA,new_AAstart,"del"),
                   T~paste0(new_REFAA,AA_START,new_VARAA))
    return(out)
    
  }else{
    return(paste0(REFAA,AA_START,VARAA))
  }
}

# function that takes the indels_summary and predicts the AA change
annotate_indels<-function(indels_summary){
  
  vcf<-VCF(
    rowRanges = GRanges("NC_045512.2",
                ranges=IRanges(
                  start=indels_summary$del_site,    end=indels_summary$del_site+nchar(indels_summary$del_from)-1)
                )
    )
  
  ref(vcf)<-DNAStringSet(indels_summary$del_from)
  alt(vcf)<-DNAStringSetList(lapply(indels_summary$del_to,function(x) c(x)))
  
  aa <- predictCoding(vcf, txdb, seqSource=fafi,if.fuzzy.codon="solve")
  
annotations<-aa%>%
   as_tibble()%>%
   unnest_wider(PROTEINLOC,names_sep=" ")%>%
   cbind(ALT2=sapply(aa$ALT,function(x) x[1]%>%as.character()))%>%
    dplyr::select(del_site=start,del_from=REF,del_to=ALT2,GENEID,REFAA,VARAA,CONSEQUENCE,AA_start=`PROTEINLOC 1`)%>%
   distinct()%>%
  rowwise()%>%
  mutate(GENEID=str_replace(GENEID,"orf","ORF"),
         AA_change=clean_AA_change(REFAA,VARAA,AA_start),
         AA_change=ifelse(is.na(AA_change),GENEID,paste0(GENEID,":",AA_change)))%>%
  ungroup()%>%
  dplyr::select(del_site,del_from,del_to,consequence=CONSEQUENCE,AA_change)%>%
   group_by(del_site,del_from,del_to,consequence)%>%
   summarise(n_options=n(),
             AA_change=paste0(AA_change,collapse="\n"))
 
  return(annotations)
}
```

Data update version: `r metadata_update_number`

Analysis last run: `r Sys.Date()`

# Aims

* Investigate the concordance of pango lineage classification between the GenBank and Viridian assemblies
* Identify and characterise indel changes between assemblies

# Results

There are n=`r metadata$Viridian_pangolin%>%unique()%>%setdiff(c("Unassigned",NA))%>%length()` pango lineages identified in the Viridian dataset, and n=`r metadata$Genbank_pangolin%>%unique()%>%setdiff(c("Unassigned",NA))%>%length()` in the Genbank dataset. There are n=`r sum(metadata$Viridian_pangolin=="Unassigned")` (`r 100*mean(metadata$Viridian_pangolin=="Unassigned")`%) sequences classified as unassigned in the Viridian assemblies versus n=`r sum(metadata$Genbank_pangolin=="Unassigned")` (`r 100*mean(metadata$Genbank_pangolin=="Unassigned")`%) in the Genbank assemblies. 

In the sections below, we first consider mismatches at the variant level, and then look more closely at the lineage level. 

## Variant level mismatches

### Overview

The plot below compares the classification of the Viridian and Genbank assemblies at the variant level, which were split out by WHO name, recombinant, unassigned lineage, or other. In general the Variant level calls are consistent - only n=`r sum(metadata$variant_same==F)` (`r 100*mean(metadata$variant_same==F)`%) samples have different variants called from the Genbank and Viridian assemblies.

```{r fig.width=6,fig.height=6}
#get limits for the marginal plots
marginal_count_limits<-c(1,max(metadata$Viridian_pangolin_variant%>%table()%>%max(),metadata$Genbank_pangolin_variant%>%table()%>%max()))

p_heatmap<-metadata%>%
  group_by(Viridian_pangolin_variant,Genbank_pangolin_variant,.drop=F)%>%
  count()%>%
  ggplot(aes(x=Viridian_pangolin_variant,y=Genbank_pangolin_variant,fill=n))+
  geom_tile()+
  geom_text(data=.%>%filter(Viridian_pangolin_variant!=Genbank_pangolin_variant,n>0),aes(label=prettyNum(n,big.mark=",")),size=2,hjust=0.5)+
  scale_x_discrete(expand=expansion(add=0.5),drop = FALSE)+
  scale_y_discrete(expand=expansion(add=0.5),drop = FALSE)+
  scale_fill_scico(palette="imola",direction=-1,trans="log10",labels=comma_format(),limits=marginal_count_limits,expand=expansion(add=0),na.value="transparent")+
  theme_bw()+
  theme(axis.text.x = element_text(hjust=0,vjust=0.5,angle=270))+
  labs(x="Viridian variant",
       y="Genbank variant",
       fill="Count")

#Add marginals
p_marginal_x<-metadata%>%
  ggplot(aes(x=Viridian_pangolin_variant,fill=after_stat(count)))+
  geom_bar()+
  scale_x_discrete(expand=expansion(add=0.5),drop = FALSE)+
  scale_y_continuous(limits=marginal_count_limits,expand=expansion(mult=c(0,0.025)),trans="log10",labels=comma_format(),breaks = trans_breaks('log10', function(x) 10^x),minor_breaks = c(1:10,1:10*10,1:10*100,1:10*1000,1:10*10000,1:10*100000,1:10*1000000))+
  scale_fill_scico(palette="imola",direction=-1,trans="log10",labels=comma_format(),limits=marginal_count_limits,expand=expansion(add=0))+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank())+
  labs(y="Count",
       fill="Count")
  
p_marginal_y<-metadata%>%
  ggplot(aes(x=Genbank_pangolin_variant,fill=after_stat(count)))+
  geom_bar()+
  scale_x_discrete(expand=expansion(add=0.5),drop = FALSE)+
  scale_y_continuous(limits=marginal_count_limits,expand=expansion(mult=c(0,0.025)),trans="log10",labels=comma_format(),breaks = trans_breaks('log10', function(x) 10^x),minor_breaks = c(1:10,1:10*10,1:10*100,1:10*1000,1:10*10000,1:10*100000,1:10*1000000))+
  scale_fill_scico(palette="imola",direction=-1,trans="log10",labels=comma_format(),limits=marginal_count_limits,expand=expansion(add=0))+
  theme_bw()+
  coord_flip()+
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_blank(),
        axis.text.x = element_text(hjust=0,vjust=0.5,angle=270))+
  labs(y="Count",
       fill="Count")

#patchwork the layout
layout<- "AB\nCD"

p_marginal_x+guide_area()+p_heatmap+p_marginal_y+plot_layout(design=layout,widths=c(1,0.5),heights=c(0.5,1),guides="collect")
```

### Association with other available metadata {.tabset}

#### By technology

```{r}
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  group_by(Platform, Viridian_amplicon_scheme)%>%
  mutate(technology_total=n())%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=variant_same))+
  geom_bar()+
  geom_text(data=.%>%select(Platform,Viridian_amplicon_scheme,technology_total)%>%distinct(),aes(x=Viridian_amplicon_scheme,y=0.0125*max(technology_total)+technology_total,label=prettyNum(technology_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.07)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Amplicon scheme",
       y="Count",
       fill="Viridian and Genbank identical")+
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=variant_same))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90),
        strip.background = element_blank(),
        strip.text=element_blank())+
  labs(x="Amplicon scheme",
       y="Proportion",
       fill="Viridian and Genbank identical")+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom")
```

#### By collection date

```{r}
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=variant_same))+
  geom_bar(width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=c(0,0.025)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Sample collection week",
       y="Count",
       fill="Viridian and Genbank identical")+
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=variant_same))+
  geom_bar(position="fill",width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom")+
  labs(x="Sample collection week",
       y="Proportion",
       fill="Viridian and Genbank identical",
       caption=paste0("n=",metadata%>%filter(!{is.na(year_week)|year_week<ymd("2020-01-01")})%>%pull(Sample)%>%length()%>%format(big.mark=",")," sequences with full collection date."))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom")

```

#### Country

```{r}
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  group_by(Country,variant_same)%>%
  count()%>%
  ungroup()%>%
  group_by(Country)%>%
  mutate(Country_total=sum(n))%>%
  ggplot(aes(x=Country,fill=variant_same,y=n))+
  geom_bar(stat="identity")+
  geom_text(data=.%>%select(Country,Country_total)%>%distinct(),aes(x=Country,y=0.01*max(Country_total)+Country_total,label=prettyNum(Country_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.06)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Country",
       y="Count",
       fill="Viridian and Genbank identical")+
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  ggplot(aes(x=Country,fill=variant_same))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90))+
  labs(x="Country",
       y="Proportion",
       fill="Viridian and Genbank identical")+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom")

```

## Lineage level

### Overall {.tabset}

Pango lineages are also generally consistent: only n=`r sum(metadata$pangolin_same==F)` (`r 100*mean(metadata$pangolin_same==F)`%) samples have different lineages called from the Genbank and Viridian assemblies. 

The heatmaps below show an overview of how the Viridian (columns) and Genbank (rows) assemblies were classified at the lineage level, coloured by a) raw count and b) scaled by the number of Viridian sequences classified as that lineage. Both axes are ordered by unaliased pango lineage, thus the boxes of clustering yellow broadly align with within-variant variation.

#### Raw count

```{r fig.height=6}
metadata%>%
  group_by(Viridian_pangolin_ordered,Viridian_pangolin_variant,Genbank_pangolin_ordered,Genbank_pangolin_variant)%>%
  count()%>%
  ggplot(aes(x=Viridian_pangolin_ordered,y=Genbank_pangolin_ordered,fill=n))+
  geom_tile()+
  coord_equal()+
  scale_x_discrete(expand=expansion(add=10),drop = FALSE)+
  scale_y_discrete(expand=expansion(add=10),drop = FALSE)+
  scale_fill_scico(palette="imola",direction=-1,trans="log10",labels=comma_format())+
  theme_bw()+
  theme(
        axis.text=element_blank(),
        axis.ticks = element_blank(),
        panel.spacing = unit(0,"cm"),
        legend.title = element_text(size=9),
        legend.position = "bottom",
        panel.background = element_rect(fill="black"),
        panel.grid = element_blank())+
  labs(x="Viridian lineage",
       y="Genbank lineage",
       fill="Count")

```

#### Scaled to Viridian call

```{r fig.height=6}
metadata%>%
  group_by(Viridian_pangolin_ordered,Viridian_pangolin_variant,Genbank_pangolin_ordered,Genbank_pangolin_variant)%>%
  count()%>%
  group_by(Viridian_pangolin_ordered,Viridian_pangolin_variant)%>%
  mutate(Viridian_lineage_n=sum(n),
         Viridian_lineage_prop=n/Viridian_lineage_n)%>%
  ggplot(aes(x=Viridian_pangolin_ordered,y=Genbank_pangolin_ordered,fill=Viridian_lineage_prop))+
  geom_tile()+
  coord_equal()+
  scale_x_discrete(expand=expansion(add=10),drop = FALSE)+
  scale_y_discrete(expand=expansion(add=10),drop = FALSE)+
  scale_fill_scico(palette="buda",direction=-1)+
  theme_bw()+
  theme(
        axis.text=element_blank(),
        axis.ticks = element_blank(),
        panel.spacing = unit(0,"cm"),
        panel.background = element_rect(fill="black"),
        panel.grid = element_blank(),
        legend.title = element_text(size=9),
        legend.position = "bottom")+
  labs(x="Viridian lineage",
       y="Genbank lineage",
       fill="Proportion of sequences relative to Viridian")
```

### Lineage relationships {.tabset}

For samples where the lineage was assigned differently between assemblies, the relationship between the two lineages was defined as:

* parent-child - i.e. B.1.1.7 and Q.1
* sibling - i.e. Q.1 and Q.2
* same variant - i.e. BA.2.13.1 and BA.5.1
* more distant - i.e. B.1.1.1 and BA.1
* one unassigned - if one of the assemblies could not be designated. 

Most mismatches were at the parent-child level (`r 100*mean(filter(metadata,pangolin_same==F)$pangolin_mismatch_relationship=="Parent-child")`%), with the Viridian assembly the child (i.e. more accurately designated) in `r 100*mean(filter(metadata,pangolin_mismatch_relationship=="Parent-child")$pangolin_mismatch_relationship_full=="Genbank Parent")`% of those.

#### All

```{r fig.height=4}
metadata%>%
  group_by(pangolin_mismatch_relationship)%>%
  count()%>%
  ggplot(aes(x=fct_rev(pangolin_mismatch_relationship),y=n,fill=pangolin_mismatch_relationship))+
  geom_bar(stat="identity")+
  geom_text(aes(y=n,label=prettyNum(n,big.mark=",")),size=2.25,hjust=-0.1)+
  scale_y_continuous(expand=expansion(mult=c(0,0.1)),labels=comma_format())+
  coord_flip()+
  scale_fill_manual(values=cols_mismatch_rel,guide="none")+
  theme_bw()+
  labs(y="Count",
       x="Relationship",
       fill="Viridian/Genbank relationship")+
  theme(legend.position="bottom",
        strip.background = element_blank(),
        strip.text=element_blank())
```

#### Mis-matched only

```{r fig.height=4}
metadata%>%
  group_by(pangolin_mismatch_relationship,pangolin_mismatch_relationship_full)%>%
  count(name="parent_total")%>%
  group_by(pangolin_mismatch_relationship)%>%
  mutate(n=sum(parent_total),
         bar_lab=ifelse(!pangolin_mismatch_relationship%in%c("Parent-child","One unassigned"),prettyNum(n,big.mark=","),paste0(prettyNum(parent_total,big.mark=","),"\n",str_replace(pangolin_mismatch_relationship_full," ","\n")))
         )%>%
  filter(pangolin_mismatch_relationship!="Same")%>%
  ggplot(aes(x=fct_rev(pangolin_mismatch_relationship),y=n,fill=pangolin_mismatch_relationship))+
  geom_bar(data=.%>%select(pangolin_mismatch_relationship,n)%>%distinct(),stat="identity",colour="black")+
  geom_bar(data=.%>%filter(pangolin_mismatch_relationship=="Parent-child"),stat="identity",aes(y=parent_total,group=fct_rev(pangolin_mismatch_relationship_full)),colour="black")+
  geom_bar(data=.%>%filter(pangolin_mismatch_relationship=="One unassigned"),stat="identity",aes(y=parent_total,group=fct_rev(pangolin_mismatch_relationship_full)),colour="black")+
  geom_text(data=.%>%filter(pangolin_mismatch_relationship%in%c("Parent-child","One unassigned")),aes(y=200+parent_total,label=bar_lab),size=2.25,hjust=0,position="stack",colour="black")+
  geom_text(data=.%>%filter(!pangolin_mismatch_relationship%in%c("Parent-child","One unassigned")),aes(y=200+n,label=bar_lab),size=2.25,hjust=0,position="stack",colour="black")+
  scale_y_continuous(expand=expansion(mult=c(0,0.1)),labels=comma_format())+
  coord_flip()+
  scale_fill_manual(values=cols_mismatch_rel,guide="none")+
  theme_bw()+
  labs(y="Count",
       x="Relationship",
       fill="Viridian/Genbank relationship")+
  theme(legend.position="bottom",
        strip.background = element_blank(),
        strip.text=element_blank())
```

### Impact of Ns {.tabset}

```{r include=F}
n_summary<-metadata%>%
  mutate(diff_in_Ns=case_when(
    Viridian_N==Genbank_N~"Equal",
    Viridian_N>Genbank_N~"Viridian higher",
    Viridian_N<Genbank_N~"Genbank higher"                         ))%>%
  group_by(diff_in_Ns)%>%
  count()%>%
  ungroup()%>%
  mutate(percent=100*n/sum(n))

metadata%>%
  filter(Viridian_pangolin_unaliased=="Unassigned")%>%
  arrange(Viridian_N)%>%
  select(Viridian_N)

metadata%>%
  filter(Genbank_pangolin_unaliased=="Unassigned")%>%
  arrange(Genbank_N)%>%
  select(Genbank_N)

metadata%>%
  mutate(diff_in_Ns=case_when(
    Viridian_N==Genbank_N~"Equal",
    Viridian_N>Genbank_N~"Viridian higher",
    Viridian_N<Genbank_N~"Genbank higher"                         ))%>%
  group_by(diff_in_Ns)%>%
  summarise(genbank_0=sum(Genbank_N==0),
            percent=100*genbank_0/n())
```
* n=`r sum(metadata$Genbank_N==0)` (`r 100*mean(metadata$Genbank_N==0)`%) of Genbank assemblies have no Ns, compared to n=`r sum(metadata$Viridian_N==0)` (`r 100*mean(metadata$Viridian_N==0)`%) of Viridian assemblies.

* n=`r filter(n_summary, diff_in_Ns=="Equal")$n` (`r filter(n_summary, diff_in_Ns=="Equal")$percent`%) of samples have same number of Ns in both assemblies (i.e. lie along y=x in the scatterplots below)
    - n=`r filter(n_summary, diff_in_Ns=="Genbank higher")$n` (`r filter(n_summary, diff_in_Ns=="Genbank higher")$percent`%) have more Ns in the Genbank assembly 
    - n=`r filter(n_summary, diff_in_Ns=="Viridian higher")$n` (`r filter(n_summary, diff_in_Ns=="Viridian higher")$percent`%) of samples have more Ns in the Viridian assembly.

#### Scatterplot

```{r}
metadata%>%
  ggplot(aes(x=Viridian_N,y=Genbank_N,colour=pangolin_mismatch_relationship))+
  geom_point(size=0.3,alpha=0)+
  geom_point(size=0.3,alpha=0.2, colour="grey30")+
  geom_abline(intercept = 0, slope = 1)+
  scale_x_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  scale_y_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  theme_bw()+
  labs(x="Ns in Viridian assembly",
       y="Ns in Genbank assembly",
       colour="Viridian/Genbank relationship")+
  guides(colour = guide_legend(nrow = 2,override.aes = list(alpha=0,size=1)))+
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.text = element_text(colour="white"),
        legend.title = element_text(size=10,colour="white"))
```

#### Scatterplot by relationship

```{r}
metadata%>%
  mutate(
    pangolin_mismatch_relationship=ifelse(Genbank_pangolin=="Unassigned"&Viridian_pangolin=="Unassigned","Both unassigned",as.character(pangolin_mismatch_relationship)),
    pangolin_mismatch_relationship=factor(pangolin_mismatch_relationship,levels=c(names(cols_mismatch_rel),"Both unassigned"))
    )%>%
  ggplot(aes(x=Viridian_N,y=Genbank_N,colour=pangolin_mismatch_relationship))+
  geom_point(size=0.3,alpha=0.3)+
  geom_abline(intercept = 0, slope = 1)+
  scale_colour_manual(values=c(cols_mismatch_rel,"Both unassigned"="Black"))+
  scale_x_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  scale_y_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  theme_bw()+
  labs(x="Ns in Viridian assembly",
       y="Ns in Genbank assembly",
       colour="Viridian/Genbank relationship")+
  guides(colour = guide_legend(nrow = 2,override.aes = list(alpha=1,size=1)))+
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))
```

#### Scatterplot, unassigned lineage only

```{r}
metadata%>%
  filter(Viridian_pangolin_variant=="Unassigned"|Genbank_pangolin_variant=="Unassigned")%>%
  mutate(pangolin_mismatch_relationship_full=ifelse(pangolin_mismatch_relationship_full=="Same","Both unassigned",pangolin_mismatch_relationship_full))%>% 
  ggplot(aes(x=Viridian_N,y=Genbank_N,colour=fct_rev(pangolin_mismatch_relationship_full)))+
  geom_point(size=0.3,alpha=0.75)+
  geom_abline(intercept = 0, slope = 1)+
  scale_x_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  scale_y_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  scale_colour_manual(values=c("Both unassigned"="Black","Viridian unassigned"="#663333","Genbank unassigned"="#225555"))+

  theme_bw()+
  labs(x="Ns in Viridian assembly",
       y="Ns in Genbank assembly",
       colour="Viridian/Genbank relationship")+
  guides(colour = guide_legend(nrow = 2,override.aes = list(alpha=1,size=1)))+
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))
```


### Association with other available metadata {.tabset}

#### By technology {.tabset}

##### Match/mis-match

```{r}
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  group_by(Platform, Viridian_amplicon_scheme)%>%
  mutate(technology_total=n())%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=pangolin_same))+
  geom_bar()+
  geom_text(data=.%>%select(Platform,Viridian_amplicon_scheme,technology_total)%>%distinct(),aes(x=Viridian_amplicon_scheme,y=0.0125*max(technology_total)+technology_total,label=prettyNum(technology_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.07)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  guides(fill = guide_legend(nrow = 2))+
  labs(x="Amplicon scheme",
       y="Count",
       fill="Viridian and Genbank identical")+
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=pangolin_same))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90),
        strip.background = element_blank(),
        strip.text=element_blank())+
  labs(x="Amplicon scheme",
       y="Proportion",
       fill="Viridian and Genbank identical")+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))
```

##### Lineage relationship

```{r}
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  group_by(Platform, Viridian_amplicon_scheme)%>%
  mutate(technology_total=n())%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=pangolin_mismatch_relationship))+
  geom_bar()+
  geom_text(data=.%>%select(Platform,Viridian_amplicon_scheme,technology_total)%>%distinct(),aes(x=Viridian_amplicon_scheme,y=0.0125*max(technology_total)+technology_total,label=prettyNum(technology_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.07)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch_rel)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Amplicon scheme",
       y="Count",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))+
metadata%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=pangolin_mismatch_relationship))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch_rel)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90),
        strip.background = element_blank(),
        strip.text=element_blank())+
  labs(x="Amplicon scheme",
       y="Proportion",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))
```

##### Lineage relationship (detail)

```{r}
metadata%>%
  filter(pangolin_mismatch_relationship!="Same")%>%
  mutate(Viridian_amplicon_scheme=str_replace(Viridian_amplicon_scheme,"^COVID-",""))%>%
  ggplot(aes(x=Viridian_amplicon_scheme,fill=pangolin_mismatch_relationship))+
  geom_bar()+
  scale_y_continuous(expand=expansion(mult=c(0,0.05)))+
  scale_fill_manual(values=cols_mismatch_rel)+
  facet_grid(.~Platform)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90),
        strip.background = element_blank(),
        #strip.text=element_blank()
        )+
  labs(x="Amplicon scheme",
       y="Count",
       fill="Viridian/Genbank relationship",
      legend.key.size = unit(0.4,"cm"))+
  guides(fill = guide_legend(nrow = 2))
```

#### By collection date {.tabset}

##### Match/mis-match

```{r}
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=pangolin_same))+
  geom_bar(width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=c(0,0.025)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Sample collection week",
       y="Count",
       fill="Viridian and Genbank identical")+
  guides(fill = guide_legend(nrow = 2))+
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=pangolin_same))+
  geom_bar(position="fill",width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom")+
  labs(x="Sample collection week",
       y="Proportion",
       fill="Viridian and Genbank identical",
       caption=paste0("n=",metadata%>%filter(!{is.na(year_week)|year_week<ymd("2020-01-01")})%>%pull(Sample)%>%length()%>%format(big.mark=",")," sequences with full collection date."))+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))

```

##### Lineage relationship

```{r}
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=pangolin_mismatch_relationship))+
  geom_bar(width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=c(0,0.025)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Sample collection week",
       y="Count",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))+
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")})%>%
  ggplot(aes(x=year_week,fill=pangolin_mismatch_relationship))+
  geom_bar(position="fill",width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom")+
  labs(x="Sample collection week",
       y="Proportion",
       fill="Viridian/Genbank relationship",
       caption=paste0("n=",metadata%>%filter(!{is.na(year_week)|year_week<ymd("2020-01-01")})%>%pull(Sample)%>%length()%>%format(big.mark=",")," sequences with full collection date."))+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))

```

##### Lineage relationship (detail)

```{r}
metadata%>%
  filter(!{is.na(year_week)|year_week<ymd("2019-01-01")},
         pangolin_mismatch_relationship!="Same")%>%
  ggplot(aes(x=year_week,fill=pangolin_mismatch_relationship))+
  geom_bar(width=6)+
  scale_x_date(expand=expansion(add=6))+
  scale_y_continuous(expand=expansion(mult=c(0,0.05)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.key.size = unit(0.4,"cm"))+
  labs(x="Sample collection week",
       y="Count",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))
```

#### Country {.tabset}

##### Match/mis-match

```{r}
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  group_by(Country,pangolin_same)%>%
  count()%>%
  ungroup()%>%
  group_by(Country)%>%
  mutate(Country_total=sum(n))%>%
  ggplot(aes(x=Country,fill=pangolin_same,y=n))+
  geom_bar(stat="identity")+
  geom_text(data=.%>%select(Country,Country_total)%>%distinct(),aes(x=Country,y=0.01*max(Country_total)+Country_total,label=prettyNum(Country_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.06)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Country",
       y="Count",
       fill="Viridian and Genbank identical")+
  guides(fill = guide_legend(nrow = 2))+
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  ggplot(aes(x=Country,fill=pangolin_same))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90))+
  labs(x="Country",
       y="Proportion",
       fill="Viridian and Genbank identical")+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))

```

##### Lineage relationship

```{r}
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  group_by(Country,pangolin_mismatch_relationship)%>%
  count()%>%
  ungroup()%>%
  group_by(Country)%>%
  mutate(Country_total=sum(n))%>%
  ggplot(aes(x=Country,fill=pangolin_mismatch_relationship,y=n))+
  geom_bar(stat="identity")+
  geom_text(data=.%>%select(Country,Country_total)%>%distinct(),aes(x=Country,y=0.01*max(Country_total)+Country_total,label=prettyNum(Country_total,big.mark=",")),size=2.25,vjust=0,inherit.aes=F)+
  scale_y_continuous(expand=expansion(mult=c(0,0.06)),labels=comma_format())+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank())+
  labs(x="Country",
       y="Count",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))+
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  ggplot(aes(x=Country,fill=pangolin_mismatch_relationship))+
  geom_bar(position="fill")+
  scale_y_continuous(expand=expansion(add=0))+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90))+
  labs(x="Country",
       y="Proportion",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))+
plot_layout(ncol=1,guides="collect")&
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.title = element_text(size=10))

```

##### Lineage relationship (detail)

```{r}
metadata%>%
  mutate(Country=str_replace(Country," ","\n"),
         Country=fct_relevel(Country,"NOT_FOUND",after=Inf))%>%
  filter(pangolin_mismatch_relationship!="Same")%>%
  ggplot(aes(x=Country,fill=pangolin_mismatch_relationship))+
  geom_bar()+
  scale_y_continuous(expand=expansion(add=0))+
  scale_x_discrete(drop=F)+
  scale_fill_manual(values=cols_mismatch_rel)+
  theme_bw()+
  theme(legend.position="bottom",
        legend.key.size = unit(0.4,"cm"),
        axis.text.x = element_text(hjust=1,vjust=0.5,angle=90))+
  labs(x="Country",
       y="Count",
       fill="Viridian/Genbank relationship")+
  guides(fill = guide_legend(nrow = 2))
```

## Indels for consistent variant calls

Indels were extracted from `all_indels.by_run.json` for samples with matching Viridian and GenBank variant-level call (or BA-lineage call for Omicron). Scatterplots show for each indel the percentage of assemblies it is identified in (with filter passed) by Viridian (x-axis) versus GenBank (y-axis); indels are labeled with site and nucleotide change if there is >5% difference between Viridian and Genbank. 

Tables give information on all indels identified in a variant, including the amino acid changes predicted from the nucleotides using `VariantAnnotation::predictCoding`.

The number of consensus deletions for variants (including all sublineages) were extracted from Cornelius Roemer's consensus .json file on [Github](https://github.com/corneliusroemer/pango-sequences) for reference, however note this does not appear to contain insertions.


```{r results='asis',fig.height=4,fig.width=8}
#initialise tibble for consensus indels
cons_indel_table<-tibble()

#loop through VOCs except Omicron
for(i in setdiff(variants_unique,c("Omicron","Recombinant","Other","Unassigned"))){
  
  cat("###", i,"\n")
  
  consensus_dels<-lapply(voc_parent[which(names(voc_parent)==i)],function(x)
    
    lapply(c(x,unlist(consensus[[x]]["children"])),function(y) consensus[[y]]["nucDeletions"]))%>%
  unlist()%>%
  setdiff("")%>%
  unique()
  
  #Get sequences for this VOC
  seqs_voc<-metadata%>%
    filter(Viridian_pangolin_variant==i,
           Genbank_pangolin_variant==i)%>%
    pull(Run)
  
  seqs_voc_indels<-tibble(sample=names(indels[seqs_voc]),data=indels[seqs_voc])%>%
  unnest_longer(data,values_to = "pipeline")%>%
  unnest_longer(pipeline,values_to="del")%>%
  rename(filter_pass=del)%>%
  pivot_wider(names_from="pipeline_id",values_from=filter_pass)

  seqs_voc_indels_summary<-seqs_voc_indels%>%
  mutate(seen_in=case_when(genbank==T&vdn==T~"both",
                           genbank==T&is.na(vdn)~"genbank",
                           vdn==T&is.na(genbank)~"viridian",
                           T~"other"))%>%
  group_by(del_id,seen_in)%>%
  count()%>%
  ungroup()%>%
  pivot_wider(names_from = "seen_in",
              values_from = n
              )
  
  if(sum(!c("del_id","both","genbank","viridian","other")%in%colnames(seqs_voc_indels_summary))>0){
    seqs_voc_indels_summary[,setdiff(c("del_id","both","genbank","viridian","other"),colnames(seqs_voc_indels_summary))]<-NA
  }
  
  seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
    mutate(viridian=ifelse(is.na(viridian),0,viridian),
        genbank=ifelse(is.na(genbank),0,genbank),
        both=ifelse(is.na(both),0,both),
        viridian=viridian+both,
        genbank=genbank+both)%>%
    filter(viridian>0|genbank>0|both>0)%>%
    mutate(across(-del_id,
                  ~{round(100*.x/length(seqs_voc),2)},
                  .names = "{.col}_percent"))%>%
    arrange(-both)%>%
    mutate(del_site=as.numeric(word(del_id,1,1,"_")),
           del_from=word(del_id,2,2,"_"),
           del_to=word(del_id,3,3,"_"),
           length=str_length(del_to)-str_length(del_from),
           inframe=ifelse(length%%3==0,TRUE,FALSE)
          )
  
    seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
      left_join(annotate_indels(seqs_voc_indels_summary))
    
    indel_prop_inframe<-seqs_voc_indels_summary%>%
      group_by(inframe)%>%
      dplyr::count()%>%
      ungroup()%>%
      mutate(prop=round(100*n/sum(n),2))
    
    indel_prop_singleton<-seqs_voc_indels_summary%>%
      mutate(singleton=
               {viridian==1&genbank==0}|
               {viridian==0&genbank==1}|
               {viridian==1&genbank==1&both==1})%>%
      group_by(singleton)%>%
      dplyr::count()%>%
      ungroup()%>%
      mutate(prop=round(100*n/sum(n),2))
  
    cat(paste0("* There are ",unique(seqs_voc_indels$del_id)%>%length()%>%format(big.mark=",")," different indels identified (filter passed) in ",seqs_voc%>%length()%>%format(big.mark=",")," sequences.\n"))
    
    cat(paste0("* n=",filter(indel_prop_singleton,singleton==TRUE)$n%>%format(big.mark = ",")," (",filter(indel_prop_singleton,singleton==TRUE)$prop,"%) indels are found only in a single sample.\n"))
        
    cat(paste0("* There are ",filter(seqs_voc_indels_summary,viridian>0,genbank==0)$del_id%>%length()%>%format(big.mark=",")," indels identified by only Viridian (of which n=",filter(seqs_voc_indels_summary,viridian==1,genbank==0)$del_id%>%length()%>%format(big.mark=",")," are found in only one sample), and ",filter(seqs_voc_indels_summary,viridian==0,genbank>0)$del_id%>%length()%>%format(big.mark=",")," by only Genbank (of which n=",filter(seqs_voc_indels_summary,viridian==0,genbank==1)$del_id%>%length()%>%format(big.mark=",")," are singleton).\n"))
    
  cat(paste0("* According to the consensus sequence, we expect up to  ",consensus_dels%>%length()," deletions in ",i, " and sublineages.\n"))
  
    cat(paste0("* We find ",filter(seqs_voc_indels_summary,both_percent>=90)$del_id%>%length()," indels in both pipelines for at least 90% of samples and ",filter(seqs_voc_indels_summary,both_percent>=50)$del_id%>%length()," indels in at least 50% of samples.\n"))
    
    cat(paste0("* ",filter(indel_prop_inframe,inframe==TRUE)$prop,"% of indels are not in frame (length not a multiple of 3).\n"))
  
    p_scatter<-seqs_voc_indels_summary%>%
      mutate(plot_lab=ifelse(
        abs(viridian_percent-genbank_percent)>5,del_id,NA))%>%
      ggplot(aes(x=viridian_percent, y=genbank_percent))+
      geom_abline(intercept=0,slope=1,colour="darkgrey")+
      geom_point(size=1,alpha=0.3)+
      ggrepel::geom_text_repel(
        aes(label=plot_lab),size=2,box.padding=0.5,min.segment.length = 0.25)+
    scale_y_continuous(limits=c(0,100))+
    scale_x_continuous(limits=c(0,100))+
    theme_bw()+
    labs(x="Viridian assemblies with indel (%)",
         y="Genbank assemblies with indel (%)",
         caption=paste0("*",unique(seqs_voc_indels$del_id)%>%length()%>%format(big.mark=",")," indels found in ",seqs_voc%>%length()%>%format(big.mark=",")," sequences"))
    
    p_hist<-seqs_voc_indels%>%
            group_by(sample)%>%
      summarise(Genbank_indel_n=sum(genbank,na.rm=T),
                Viridian_indel_n=sum(vdn,na.rm=T))%>%
      full_join(data.frame(sample=seqs_voc))%>%
      mutate(across(ends_with("_n"),coalesce,0))%>%
      pivot_longer(cols=ends_with("_n"),names_to = "assembly",values_to="indel_n")%>%
      mutate(assembly=str_replace(assembly,"_indel_n",""))%>%
      ggplot(aes(x=indel_n,fill=assembly))+
      geom_histogram(binwidth = 1,stat="count",position=position_dodge2(width = 0.9, preserve = "single"),alpha=0.9)+
      geom_vline(xintercept=length(consensus_dels),linetype="dashed",colour="grey")+
      stat_summary(aes(x = 1, y = indel_n, xintercept = stat(y),colour=assembly,linetype=assembly),
               fun.y = median, geom = "vline") +
      scale_x_continuous(expand=expansion(add=1),label = ~ scales::comma(.x, accuracy = 1))+
      scale_y_continuous(labels=comma_format(),expand=expansion(add=0))+
      scale_colour_manual(values=cols_assembly_dark,guide="none")+
      scale_fill_manual(values=cols_assembly)+
      scale_linetype(guide="none")+
  theme_bw()+
  theme(legend.position=c(1,1),
        legend.justification = c(1,1),
        legend.key.size = unit(0.25,"cm"),
        legend.background = element_blank(),
        legend.title = element_blank())+
      labs(x="Number of indels identified in sample",
           y="Number of samples",
           caption="Grey dashed line shows deletions in consensus")
    
    print(p_scatter+p_hist+plot_layout(nrow=1))
    cat("\n\n")
    
    cons_indels<-seqs_voc_indels_summary%>%
      filter(both_percent>50)%>%
      select(del_id,both_percent)%>%
      mutate(variant=i)
    
    cons_indel_table<-cons_indel_table%>%
      bind_rows(cons_indels)
    
    # finalise summary for data table
    
    seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
    mutate(
      del_from=ifelse(abs(nchar(del_from))>12,paste0(str_sub(del_from,1,12),"..."),del_from),
      del_to=ifelse(abs(nchar(del_to))>12,paste0(str_sub(del_to,1,12),"..."),del_to))%>%
      dplyr::select(
        `Nt site`=del_site,
        `Ref nt`=del_from,
        `Alt nt`=del_to,
        `AA change`=AA_change,
        `Indel length`=length,
        `In frame`=inframe,
        Consequence=consequence,
        `Viridian\n(n)` = viridian,
        `Viridian\n(%)` = viridian_percent,
        `Genbank\n(n)` = genbank,
        `Genbank\n(%)` = genbank_percent,
        `Both\n(n)` = both,
        `Both\n(%)` = both_percent)
    

    sketch <- withTags(
      table(
        class = "display",
        thead(
          tr(
            th(colspan = 7, ""),
            th(colspan = 2, "Viridian", style = "border-left: solid 0.5px;"),
            th(colspan = 2, "Genbank", style = "border-left: solid 0.5px;"),
            th(colspan = 2, "Both", style = "border-left: solid 0.5px;")
            ),
          tr(
            lapply(colnames(seqs_voc_indels_summary)%>%str_replace("Viridian\n","")%>%str_replace("Genbank\n","")%>%str_replace("Both\n",""),
                   function(x){
                     if(x=="(n)"){
                       th(x,style = "border-left: solid 1px;")
                       }else{
                         th(x)
                         }
                     }
                   )
            )
          )
        )
      )
    
    seqs_voc_indels_summary%>%
      DT::datatable(
        container = sketch,
        rownames = FALSE,
        escape = FALSE,
        class="compact",
        extensions="FixedColumns",
        options = list(pageLength = 10,
                       autoWidth = TRUE,
                       lengthChange=FALSE,
                       scrollX = '400px',
                       scroller = TRUE,
                       fixedColumns = list(leftColumns = 3)
                       ))%>%
    formatRound(c(8,10,12), digits = 0)%>%
    formatStyle(c(8,10,12), `border-left` = "solid 1px")%>%
    knitr::knit_print() %>%
    cat()
  
  cat("\n\n")
  
}

```

### Omicron {.tabset}

```{r results='asis',fig.height=4,fig.width=8}

metadata_omicron<-metadata%>%
  filter(Viridian_pangolin_variant=="Omicron",
         Genbank_pangolin_variant=="Omicron")%>%
  mutate(Viridian_pangolin_BA=word(Viridian_pangolin_unaliased,1,5,sep="\\."),
         Genbank_pangolin_BA=word(Genbank_pangolin_unaliased,1,5,sep="\\."))%>%
  filter(Viridian_pangolin_BA==Genbank_pangolin_BA)

for(i in str_sort(unique(metadata_omicron$Viridian_pangolin_BA),numeric=T)){
  
  cat("####", str_replace(i,"B.1.1.529","BA"),"\n")
  
  consensus_dels<-lapply(paste0("BA.",word(i,-1,-1,"\\.")),function(x)
    
    lapply(c(x,unlist(consensus[[x]]["children"])),function(y) consensus[[y]]["nucDeletions"]))%>%
  unlist()%>%
  setdiff("")%>%
  unique()
    
  #Get sequences for this VOC
  seqs_voc<-metadata_omicron%>%
    filter(Viridian_pangolin_BA==i,
           Genbank_pangolin_BA==i)%>%
    pull(Run)
  
  seqs_voc_indels<-tibble(sample=names(indels[seqs_voc]),data=indels[seqs_voc])%>%
  unnest_longer(data,values_to = "pipeline")%>%
  unnest_longer(pipeline,values_to="del")%>%
  dplyr::rename(filter_pass=del)%>%
  pivot_wider(names_from="pipeline_id",values_from=filter_pass)

  seqs_voc_indels_summary<-seqs_voc_indels%>%
  mutate(seen_in=case_when(genbank==T&vdn==T~"both",
                           genbank==T&is.na(vdn)~"genbank",
                           vdn==T&is.na(genbank)~"viridian",
                           T~"other"))%>%
  group_by(del_id,seen_in)%>%
  dplyr::count()%>%
  ungroup()%>%
  pivot_wider(names_from = "seen_in",
              values_from = n
              #values_from = prop
              )
  
  if(sum(!c("del_id","both","genbank","viridian","other")%in%colnames(seqs_voc_indels_summary))>0){
    seqs_voc_indels_summary[,setdiff(c("del_id","both","genbank","viridian","other"),colnames(seqs_voc_indels_summary))]<-NA
  }
  
  seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
    mutate(viridian=ifelse(is.na(viridian),0,viridian),
        genbank=ifelse(is.na(genbank),0,genbank),
        both=ifelse(is.na(both),0,both),
        viridian=viridian+both,
        genbank=genbank+both)%>%
    filter(viridian>0|genbank>0|both>0)%>%
    mutate(across(-del_id,
                  ~{round(100*.x/length(seqs_voc),2)},
                  .names = "{.col}_percent"))%>%
    arrange(-both)%>%
    mutate(del_site=as.numeric(word(del_id,1,1,"_")),
           del_from=word(del_id,2,2,"_"),
           del_to=word(del_id,3,3,"_"),
           length=str_length(del_to)-str_length(del_from),
           inframe=ifelse(length%%3==0,TRUE,FALSE)
          )
  
    seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
      left_join(annotate_indels(seqs_voc_indels_summary))
    
    indel_prop_inframe<-seqs_voc_indels_summary%>%
      group_by(inframe)%>%
      dplyr::count()%>%
      ungroup()%>%
      mutate(prop=round(100*n/sum(n),2))
    
    indel_prop_singleton<-seqs_voc_indels_summary%>%
      mutate(singleton=
               {viridian==1&genbank==0}|
               {viridian==0&genbank==1}|
               {viridian==1&genbank==1&both==1})%>%
      group_by(singleton)%>%
      dplyr::count()%>%
      ungroup()%>%
      mutate(prop=round(100*n/sum(n),2))
  
    cat(paste0("* There are ",unique(seqs_voc_indels$del_id)%>%length()%>%format(big.mark=",")," different indels identified in ",seqs_voc%>%length()%>%format(big.mark=",")," sequences.\n"))
  cat(paste0("* According to the consensus sequence, we expect  up to  ",consensus_dels%>%length()," deletions in ",str_replace(i,"B.1.1.529","BA"), " and sublineages.\n"))
  
    cat(paste0("* n=",filter(indel_prop_singleton,singleton==TRUE)$n%>%format(big.mark = ",")," (",filter(indel_prop_singleton,singleton==TRUE)$prop,"%) indels are found only in a single sample.\n"))
    
    cat(paste0("* There are ",filter(seqs_voc_indels_summary,viridian>0,genbank==0)$del_id%>%length()%>%format(big.mark=",")," indels identified by only Viridian (of which n=",filter(seqs_voc_indels_summary,viridian==1,genbank==0)$del_id%>%length()%>%format(big.mark=",")," are found in only one sample), and ",filter(seqs_voc_indels_summary,viridian==0,genbank>0)$del_id%>%length()%>%format(big.mark=",")," by only Genbank (of which n=",filter(seqs_voc_indels_summary,viridian==0,genbank==1)$del_id%>%length()%>%format(big.mark=",")," are singleton).\n"))
    
    cat(paste0("* We find ",filter(seqs_voc_indels_summary,both_percent>=90)$del_id%>%length()," indels in both pipelines for at least 90% of samples and ",filter(seqs_voc_indels_summary,both_percent>=50)$del_id%>%length()," indels in at least 50% of samples.\n"))
    
    cat(paste0("* ",filter(indel_prop_inframe,inframe==TRUE)$prop,"% of indels are not in frame (length not a multiple of 3).\n"))
  
    p_scatter<-seqs_voc_indels_summary%>%
      mutate(plot_lab=ifelse(#{both_percent>5&both_percent<95}|
            abs(viridian_percent-genbank_percent)>5,del_id,NA))%>%
      ggplot(aes(x=viridian_percent, y=genbank_percent))+
      geom_abline(intercept=0,slope=1,colour="darkgrey")+
      geom_point(size=1,alpha=0.3)+
      ggrepel::geom_text_repel(
        aes(label=plot_lab),size=2,box.padding=0.5,min.segment.length = 0.25)+
    scale_y_continuous(limits=c(0,100))+
    scale_x_continuous(limits=c(0,100))+
    theme_bw()+
    labs(x="Viridian assemblies with indel (%)",
         y="Genbank assemblies with indel (%)",
         caption=paste0("*",unique(seqs_voc_indels$del_id)%>%length()%>%format(big.mark=",")," indels found in ",seqs_voc%>%length()%>%format(big.mark=",")," sequences"))
    
    p_hist<-seqs_voc_indels%>%
            group_by(sample)%>%
      summarise(Genbank_indel_n=sum(genbank,na.rm=T),
                Viridian_indel_n=sum(vdn,na.rm=T))%>%
      full_join(data.frame(sample=seqs_voc))%>%
      mutate(across(ends_with("_n"),coalesce,0))%>%
      pivot_longer(cols=ends_with("_n"),names_to = "assembly",values_to="indel_n")%>%
      mutate(assembly=str_replace(assembly,"_indel_n",""))%>%
      ggplot(aes(x=indel_n,fill=assembly))+
      geom_histogram(binwidth = 1,stat="count",position=position_dodge2(width = 0.9, preserve = "single"),alpha=0.9)+
      geom_vline(xintercept=length(consensus_dels),linetype="dashed",colour="grey")+
      stat_summary(aes(x = 1, y = indel_n, xintercept = stat(y),colour=assembly,linetype=assembly),
               fun.y = median, geom = "vline") +
      scale_x_continuous(expand=expansion(add=1),label = ~ scales::comma(.x, accuracy = 1))+
      scale_y_continuous(labels=comma_format(),expand=expansion(add=0))+
      scale_colour_manual(values=cols_assembly_dark,guide="none")+
      scale_fill_manual(values=cols_assembly)+
      scale_linetype(guide="none")+
  theme_bw()+
  theme(legend.position=c(1,1),
        legend.justification = c(1,1),
        legend.key.size = unit(0.25,"cm"),
        legend.background = element_blank(),
        legend.title = element_blank())+
      labs(x="Number of indels identified in sample",
           y="Number of samples",
           caption="Grey dashed line shows deletions in consensus")
    
    print(p_scatter+p_hist+plot_layout(nrow=1))
    cat("\n\n")
    
    cons_indels<-seqs_voc_indels_summary%>%
      filter(both_percent>50)%>%
      select(del_id,both_percent)%>%
      mutate(variant=paste("Omicron",str_replace(i,"B.1.1.529","BA")))
    
    cons_indel_table<-cons_indel_table%>%
      bind_rows(cons_indels)
    
    # finalise table for data summary
    seqs_voc_indels_summary<-seqs_voc_indels_summary%>%
    mutate(
      del_from=ifelse(abs(nchar(del_from))>12,paste0(str_sub(del_from,1,12),"..."),del_from),
      del_to=ifelse(abs(nchar(del_to))>12,paste0(str_sub(del_to,1,12),"..."),del_to))%>%
      dplyr::select(
        `Nt site`=del_site,
        `Ref nt`=del_from,
        `Alt nt`=del_to,
        `AA change`=AA_change,
        `Indel length`=length,
        `In frame`=inframe,
        Consequence=consequence,
        `Viridian\n(n)` = viridian,
        `Viridian\n(%)` = viridian_percent,
        `Genbank\n(n)` = genbank,
        `Genbank\n(%)` = genbank_percent,
        `Both\n(n)` = both,
        `Both\n(%)` = both_percent)
    

    sketch <- withTags(
      table(
        class = "display",
        thead(
          tr(
            th(colspan = 7, ""),
            th(colspan = 2, "Viridian", style = "border-left: solid 0.5px;"),
            th(colspan = 2, "Genbank", style = "border-left: solid 0.5px;"),
            th(colspan = 2, "Both", style = "border-left: solid 0.5px;")
            ),
          tr(
            lapply(colnames(seqs_voc_indels_summary)%>%str_replace("Viridian\n","")%>%str_replace("Genbank\n","")%>%str_replace("Both\n",""),
                   function(x){
                     if(x=="(n)"){
                       th(x,style = "border-left: solid 1px;")
                       }else{
                         th(x)
                         }
                     }
                   )
            )
          )
        )
      )
    
    seqs_voc_indels_summary%>%
      DT::datatable(
        container = sketch,
        rownames = FALSE,
        escape = FALSE,
        class="compact",
        extensions="FixedColumns",
        options = list(pageLength = 10,
                       autoWidth = TRUE,
                       lengthChange=FALSE,
                       scrollX = '400px',
                       scroller = TRUE,
                       fixedColumns = list(leftColumns = 3)
                       ))%>%
    formatRound(c(8,10,12), digits = 0)%>%
    formatStyle(c(8,10,12), `border-left` = "solid 1px")%>%
    knitr::knit_print() %>%
    cat()
  
  cat("\n\n")

}

```

### Indels for mis-matched variant calls

```{r}
#Get sequences where variant calls are mismatched
seqs_mismatchvar<-metadata%>%
  filter(variant_same==F,
         !{Genbank_pangolin_variant=="Unassigned"|Viridian_pangolin_variant=="Unassigned"})%>%
    pull(Run)
```

There are `r length(seqs_mismatchvar)` sequences where a WHO variant has been assigned to both the Viridian and Genbank assemblies but does not match. Repeating the previous indel analysis summarised across all the sequences with variant mismatch, we see a subset of indels found in more Viridian assemblies than Genbank (left plot). These indels also overlap with the Variant consensus indels from WHO variants (highlighted in red, defined as found in >50% samples by both Viridian and Genbank in previous variant analysis). We also see that Genbank assemblies have fewer indels called per sample (right plot; dashed lines give medians).

```{r fig.height=4,fig.width=8}
#Get the indels
seqs_mismatchvar_indels<-tibble(sample=names(indels[seqs_mismatchvar]),data=indels[seqs_mismatchvar])%>%
  unnest_longer(data,values_to = "pipeline")%>%
  unnest_longer(pipeline,values_to="del")%>%
  rename(filter_pass=del)%>%
  pivot_wider(names_from="pipeline_id",values_from=filter_pass)

#Summarise indels
seqs_mismatchvar_indels_summary<-seqs_mismatchvar_indels%>%
    full_join(seqs_mismatchvar_indels%>%tidyr::expand(sample,del_id))%>%
    mutate(seen_in=case_when(genbank==T&vdn==T~"both",
                           genbank==T&is.na(vdn)~"genbank",
                           vdn==T&is.na(genbank)~"viridian",
                           T~"missing"))%>%
  group_by(del_id,seen_in)%>%
  count()%>%
  ungroup()%>%
  group_by(del_id)%>%
  mutate(percent=100*n/sum(n))%>%
  pivot_wider(names_from = "seen_in",
              values_from = c("n","percent"),
              values_fill=0
              )%>%
  mutate(
        n_viridian=n_viridian+n_both,
        n_genbank=n_genbank+n_both,
        percent_viridian=percent_viridian+percent_both,
        percent_genbank=percent_genbank+percent_both,
        )

seqs_mismatchvar_indels_summary%>%
  mutate(plot_lab=ifelse(abs(percent_viridian-percent_genbank)>5,del_id,NA))%>%
  ggplot(aes(x=percent_viridian,y=percent_genbank))+
  geom_abline(intercept=0,slope=1,colour="darkgrey")+
  geom_point()+
  geom_point(data=.%>%filter(del_id%in%cons_indel_table$del_id),colour="red")+
  ggrepel::geom_text_repel(
        aes(label=plot_lab),size=2)+
  scale_y_continuous(limits=c(0,100))+
    scale_x_continuous(limits=c(0,100))+
    theme_bw()+
  labs(x="Viridian assemblies with indel (%)",
       y="Genbank assemblies with indel (%)")+

seqs_mismatchvar_indels%>%
      group_by(sample)%>%
      summarise(Genbank_indel_n=sum(genbank,na.rm=T),
                Viridian_indel_n=sum(vdn,na.rm=T))%>%
      full_join(data.frame(sample=seqs_mismatchvar))%>%
      mutate(across(ends_with("_n"),coalesce,0))%>%
      pivot_longer(cols=ends_with("_n"),names_to = "assembly",values_to="indel_n")%>%
      mutate(assembly=str_replace(assembly,"_indel_n",""))%>%
      ggplot(aes(x=indel_n,fill=assembly))+
      geom_histogram(binwidth = 1,stat="count",position=position_dodge2(width = 0.9, preserve = "single"),alpha=0.9)+
      stat_summary(aes(x = 1, y = indel_n, xintercept = stat(y),colour=assembly,linetype=assembly),
               fun.y = median, geom = "vline") +
      scale_x_continuous(expand=expansion(add=1),label = ~ scales::comma(.x, accuracy = 1))+
      scale_y_continuous(labels=comma_format(),expand=expansion(add=0))+
      scale_colour_manual(values=cols_assembly_dark,guide="none")+
      scale_fill_manual(values=cols_assembly)+
      scale_linetype(guide="none")+
  theme_bw()+
  theme(legend.position=c(1,1),
        legend.justification = c(1,1),
        legend.key.size = unit(0.25,"cm"),
        legend.background = element_blank(),
        legend.title = element_blank())+
      labs(x="Number of indels identified in sample",
           y="Number of samples",
           caption="Grey dashed line shows deletions in consensus")+
  plot_layout(nrow=1)
```

To investigate further, we considered all pairs of Viridian-Genbank WHO variant mismatches. For each Variant consensus indel, we compared the number of samples where it was not identified (left of black line) to that where it was (right of black line) using Viridian (blue) and Genbank (red). The purple bar overlap shows where the presence/absence is consistent between the two assemblies. The WHO variants in which the indel is consensus are listed under the site identifier.

The biggest discrepancies are seen between Delta and Omicron consensus indels, where variant-defining indel presence/absence aligns with the Variant call. 

```{r fig.height=14,fig.width=8}

cons_indel_table_col<-cons_indel_table%>%
  group_by(del_id)%>%
  summarise(lab=paste0(str_replace(variant,"Omicron ",""),collapse=", "))

seqs_mismatchvar_indels%>%
  filter(del_id%in%cons_indel_table$del_id)%>%
  complete(del_id,sample=seqs_mismatchvar)%>%
  complete(sample,del_id=cons_indel_table$del_id)%>%
  left_join(metadata,by=c("sample"="Run"))%>%
  group_by(Viridian_pangolin_variant,Genbank_pangolin_variant,del_id)%>%
  summarise(total=n(),
            genbank_present=sum(genbank==T,na.rm=T),
            genbank_absent=sum(genbank!=T|is.na(genbank),na.rm=T),
            viridian_present=sum(vdn==T,na.rm=T),
            viridian_absent=sum(vdn!=T|is.na(vdn),na.rm=T))%>%
  pivot_longer(cols=c(ends_with("present"),ends_with("absent")),
               names_to="assembly_seen",
               values_to="count")%>%
  left_join(cons_indel_table_col)%>%
  mutate(
    assembly=word(assembly_seen,1,1,"_")%>%str_to_title(),
    n_plot=ifelse(str_detect(assembly_seen,"absent"),-1*count,count),
    variant_lab=paste0("<span style = 'color:blue;font-size:7pt'>",Viridian_pangolin_variant,"</span>-<span style = 'color:red;font-size:7pt'>",Genbank_pangolin_variant,"</span>"),
    site=as.numeric(word(del_id,1,1,"_")),
    facet_lab=paste0(site,"<br><span style='font-size:7pt'>",lab,"</span>"),
    facet_lab=fct_reorder(facet_lab,site)
  )%>%
  ggplot(aes(x=variant_lab,y=n_plot,fill=assembly))+
  geom_bar(stat="identity",position="identity",alpha=0.5)+
  geom_hline(yintercept=0,colour="black")+
  scale_fill_manual(values=c("Viridian"="blue","Genbank"="red"))+
  scale_y_continuous(labels=~abs(.x),
                     expand=expansion(add=2),
                     sec.axis = sec_axis(
                       transform = ~.x/max(.x),
                       breaks=c(-0.5,0.5),
                       labels=c("absent","present"))
                    )+
  coord_flip()+
  facet_wrap(.~facet_lab,ncol=4)+
  theme_bw()+
  theme(
    legend.position = "bottom",
    legend.key.size = unit(0.25,"cm"),
    axis.text.y = element_markdown(),
    axis.title.y = element_markdown(),
    axis.ticks.x.top = element_blank(),
    strip.background = element_blank(),
    strip.text.x = element_textbox(
      width = unit(1, "npc"),
      halign=0.5,
      height = unit(2,"lines"),
      padding = margin(0, 0, 0, 0),
      margin = margin(0.5,0.5, 0.5, 0.5),
      valign = 0),
    strip.placement = "outside")+
  labs(y="Count",
       x="WHO variant mismatch (<span style = 'color:blue;'>Viridian</span>-<span style = 'color:red;'>Genbank</span>)",
       fill="Assembly")
  
```

```{r include=F}
#Code for figures

# Ns
metadata%>%
  ggplot(aes(x=Viridian_N,y=Genbank_N,colour=pangolin_mismatch_relationship))+
  geom_point(size=0.3,alpha=0)+
  geom_point(size=0.3,alpha=0.2, colour="grey30")+
  geom_abline(intercept = 0, slope = 1)+
  scale_x_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  scale_y_continuous(limits=c(0,NA),expand = expansion(mult=0.025),
                     labels=comma_format())+
  theme_bw()+
  labs(x="Ns in Viridian assembly",
       y="Ns in Genbank assembly",
       colour="Viridian/Genbank relationship")+
  guides(colour = guide_legend(nrow = 2,override.aes = list(alpha=0,size=1)))+
  theme(legend.position = "bottom",
        legend.key.size = unit(0.4,"cm"),
        legend.text = element_text(colour="white"),
        legend.title = element_text(size=10,colour="white"))

ggsave("2024-04_Ns.png",width=6,height=5,unit="in",dpi=300)
ggsave("2024-04_Ns.pdf",width=6,height=5,unit="in",dpi=300)

#Extracted indel plot - only include indels with differences 
seqs_mismatchvar_indels%>%
  filter(del_id%in%cons_indel_table$del_id)%>%
  complete(del_id,sample=seqs_mismatchvar)%>%
  complete(sample,del_id=cons_indel_table$del_id)%>%
  left_join(metadata,by=c("sample"="Run"))%>%
  group_by(Viridian_pangolin_variant,Genbank_pangolin_variant,del_id)%>%
  summarise(total=n(),
            genbank_present=sum(genbank==T,na.rm=T),
            genbank_absent=sum(genbank!=T|is.na(genbank),na.rm=T),
            viridian_present=sum(vdn==T,na.rm=T),
            viridian_absent=sum(vdn!=T|is.na(vdn),na.rm=T))%>%
  mutate(both_same=viridian_present==genbank_present)%>%
  group_by(del_id)%>%
  mutate(any_diff=n()-sum(both_same))%>%
  filter(any_diff>0)%>%
  pivot_longer(cols=c(ends_with("present"),ends_with("absent")),
               names_to="assembly_seen",
               values_to="count")%>%
  left_join(cons_indel_table_col)%>%
  mutate(
    assembly=word(assembly_seen,1,1,"_")%>%str_to_title(),
    n_plot=ifelse(str_detect(assembly_seen,"absent"),-1*count,count),
    variant_lab=paste0("<span style = 'color:blue;font-size:6pt'>",Viridian_pangolin_variant,"</span>-<span style = 'color:red;font-size:6pt'>",Genbank_pangolin_variant,"</span>"),
    site=as.numeric(word(del_id,1,1,"_")),
    facet_lab=paste0(site,"<br><span style='font-size:6pt'>",lab,"</span>"),
    facet_lab=fct_reorder(facet_lab,site)
  )%>%
  ggplot(aes(x=variant_lab,y=n_plot,fill=assembly))+
  geom_bar(stat="identity",position="identity",alpha=0.5,width=0.8)+
  geom_hline(yintercept=0,colour="black",linewidth=0.1)+
  scale_fill_manual(values=c("Viridian"="blue","Genbank"="red"))+
  scale_y_continuous(labels=~abs(.x),
                     expand=expansion(add=2),
                     sec.axis = sec_axis(
                       transform = ~.x/max(.x),
                       breaks=c(-0.5,0.5),
                       labels=c("absent","present"))
                    )+
  coord_flip()+
  facet_wrap(.~fct_reorder(facet_lab,site),ncol=5)+
  theme_bw(base_size=7)+
  theme(
    legend.position = "bottom",
    legend.key.size = unit(0.25,"cm"),
    axis.text.y = element_markdown(),
    axis.title.y = element_markdown(),
    axis.ticks.x.top = element_blank(),
    strip.background = element_blank(),
    strip.text.x = element_textbox(
      width = unit(1, "npc"),
      halign=0.5,
      height = unit(1.8,"lines"),
      padding = margin(0, 0, 0, 0),
      margin = margin(0.5,0.5, 0.5, 0.5),
      valign = 0),
    strip.placement = "outside")+
  labs(y="Count",
       x="WHO variant mismatch (<span style = 'color:blue;'>Viridian</span>-<span style = 'color:red;'>Genbank</span>)",
       fill="Assembly")

ggsave("2024-04_indel_mismatch.pdf",width=6,height=6,unit="in")
ggsave("2024-04_indel_mismatch.png",width=6,height=6,unit="in",dpi=300)
```

